src/Microsoft.Data.SqlClient/add-ons/AzureKeyVaultProvider/AzureSqlKeyCryptographer.cs

```diff
@@ -0,0 +1,233 @@
+        /// <returns></returns>
+        internal KeyVaultKey GetKey(string keyIdentifierUri)
+        {
+            if (_keyDictionary.ContainsKey(keyIdentifierUri))
+            {
+                return _keyDictionary[keyIdentifierUri];
+            }
+
+            if (_keyFetchTaskDictionary.ContainsKey(keyIdentifierUri))
+            {
+                return Task.Run(() => _keyFetchTaskDictionary[keyIdentifierUri]).Result;
+            }
+
+            // Not a public exception - not likely to occur.
+            throw new KeyNotFoundException($"The key with identifier {keyIdentifierUri} was not found.");
+        }
+
+        /// <summary>
+        /// Gets the public Key size in bytes.
+        /// </summary>
+        /// <param name="keyIdentifierUri">The key vault key identifier URI</param>
+        /// <returns></returns>
+        internal int GetKeySize(string keyIdentifierUri)
+        {
+            return GetKey(keyIdentifierUri).Key.N.Length;
+        }
+
+        /// <summary>
+        /// Generates signature based on RSA PKCS#v1.5 scheme using a specified Azure Key Vault Key URL. 
+        /// </summary>
+        /// <param name="message">The data to sign</param>
+        /// <param name="keyIdentifierUri">The key vault key identifier URI</param>
+        /// <returns></returns>
+        internal byte[] SignData(byte[] message, string keyIdentifierUri)
+        {
+            CryptographyClient cryptographyClient = GetCryptographyClient(keyIdentifierUri);
+            return cryptographyClient.SignData(RS256, message).Signature;
+        }
+
+        internal bool VerifyData(byte[] message, byte[] signature, string keyIdentifierUri)
+        {
+            CryptographyClient cryptographyClient = GetCryptographyClient(keyIdentifierUri);
+            return cryptographyClient.VerifyData(RS256, message, signature).IsValid;
+        }
+
+        internal byte[] UnwrapKey(KeyWrapAlgorithm keyWrapAlgorithm, byte[] encryptedKey, string keyIdentifierUri)
+        {
+            CryptographyClient cryptographyClient = GetCryptographyClient(keyIdentifierUri);
+            return cryptographyClient.UnwrapKey(keyWrapAlgorithm, encryptedKey).Key;
+        }
+
+        internal byte[] WrapKey(KeyWrapAlgorithm keyWrapAlgorithm, byte[] key, string keyIdentifierUri)
+        {
+            CryptographyClient cryptographyClient = GetCryptographyClient(keyIdentifierUri);
+            return cryptographyClient.WrapKey(keyWrapAlgorithm, key).EncryptedKey;
+        }
+
+        private CryptographyClient GetCryptographyClient(string keyIdentifierUri)
+        {
+            if (cryptoClientDictionary.ContainsKey(keyIdentifierUri))
+            {
+                return cryptoClientDictionary[keyIdentifierUri];
+            }
+
+            CryptographyClient cryptographyClient = new CryptographyClient(GetKey(keyIdentifierUri).Id, TokenCredential);
+            cryptoClientDictionary[keyIdentifierUri] = cryptographyClient;
```

Comment:
There's an atomic `AddOrUpdate` method that may yield better performance and less potential overhead (you may allocate more here than you need to with concurrent calls).